consensus/src/pipeline/virtual_processor/processor.rs

```diff
@@ -1188,6 +1191,190 @@ impl VirtualStateProcessor {
         Ok(())
     }
 
+    pub fn get_utxo_return_address(
+        &self,
+        txid: Hash,
+        target_daa_score: u64,
+        source_hash: Hash,
+        config: &Config,
+    ) -> Result<Address, ReturnAddressError> {
+        // We need consistency between the utxo_diffs_store, block_transactions_store, selected chain and header store reads
+        let guard = self.pruning_lock.blocking_read();
+
+        let source_daa_score = self
+            .headers_store
+            .get_compact_header_data(source_hash)
+            .map(|compact_header| compact_header.daa_score)
+            .map_err(|_| ReturnAddressError::NotFound(format!("Did not find compact header for source hash {}", source_hash)))?;
+
+        if target_daa_score < source_daa_score {
+            // Early exit if target daa score is lower than that of pruning point's daa score:
+            return Err(ReturnAddressError::AlreadyPruned);
+        }
+
+        let (matching_chain_block_hash, acceptance_data) =
+            self.find_accepting_chain_block_hash_at_daa_score(target_daa_score, source_hash, guard)?;
+
+        let (index, containing_acceptance) = self
+            .find_tx_acceptance_data_and_index_from_block_acceptance(txid, acceptance_data.clone())
+            .ok_or(ReturnAddressError::NotFound(format!("Did not find containing_acceptance for tx {}", txid)))?;
+
+        // Found Merged block containing the TXID
+        let tx = self
+            .block_transactions_store
+            .get(containing_acceptance.block_hash)
+            .map_err(|_| ReturnAddressError::NotFound(format!("Did not block {} at block tx store", containing_acceptance.block_hash)))
+            .and_then(|block_txs| {
+                block_txs.get(index).cloned().ok_or_else(|| {
+                    ReturnAddressError::NotFound(format!(
```

Comment:
Since you already have a dedicated error type, I would make an err variant for each and every err instance in this method and completely avoid inline string descriptions. It will significantly improve readability